lib/LanguagePicker.js

```diff
@@ -19,50 +42,86 @@ function LanguagePicker(lang = 'en', config = {}) {
   // Use the given language as first choice
   fallbacks = [lang].concat(fallbacks);
 
-  // Fallback to English
-  if (fallbacks.indexOf('en') === -1) {
-    fallbacks.push('en');
-  }
-
-  // Fallback to nameTag
-  if (this.nameTag) {
-    fallbacks.push(this.nameTag);
-  }
+  // Remove duplicates
+  fallbacks = fallbacks.filter((item, i) => fallbacks.indexOf(item) === i);
 
   // Add prefix to all languages if exists
   // eslint-disable-next-line arrow-body-style
   fallbacks = fallbacks.map((code) => {
     return code === this.nameTag ? code : this.prefix + code;
   });
 
+  // Store initial fallbacks
   this.fallbacks = fallbacks;
 }
 
 LanguagePicker.prototype.newProcessor = function newProcessor() {
-  const { fallbacks, nameTag, prefix } = this;
-  let bestChoiceLang = nameTag || (prefix ? `${prefix}en` : 'en');
-  let bestChoiceValue;
   let firstChoiceValue;
+  const prefixedEnglish = this.prefix ? `${this.prefix}en` : 'en';
+  const prefixedLangScript = `-${this.langScript}`;
 
   return {
-    addValue(lang, value) {
-      const langIndex = fallbacks.indexOf(lang);
-      if (
-        langIndex > -1 &&
-        langIndex <= fallbacks.indexOf(bestChoiceLang)
-      ) {
-        // Given language is a better fallback
-        bestChoiceLang = lang;
-        bestChoiceValue = value;
+    addValue: (lang, value) => {
+      this.values[lang] = value;
+
+      if (!firstChoiceValue) {
+        firstChoiceValue = value;
+      }
+    },
+    getResult: () => {
+      let val;
+
+      // Get the best value from the best language fallback:
+      // 1. Requested language
+      val = this.values[this.userLang];
+      if (val) {
+        return val;
+      }
+
+      // 2. Fallback language from fallbacks.json
+      for (const fallback of this.fallbacks) {
+        val = this.values[fallback];
+        if (val) {
+          return val;
+        }
+      }
+
+      // 3. Any language with suffix of same script as requested language
+      const valueLangCodes = Object.keys(this.values);
+      for (const langCode of valueLangCodes) {
+        if (langCode.endsWith(prefixedLangScript)) {
+          return this.values[langCode];
+        }
+      }
+
+      // 4. Another language with the same script
+      for (const langCode of this.languagesInScript) {
+        val = this.values[langCode];
+        if (val) {
+          return val;
+        }
+      }
+
+      // 5. English
+      val = this.values[prefixedEnglish];
+      if (val) {
+        return val;
+      }
+
+      // 6. Any language with suffix 'Latn'
+      // Only do this if the script isn't already latin;
+      // otherwise, we've done it in step 3
+      for (const langCode of valueLangCodes) {
+        if (langCode.endsWith('-Latn')) {
+          return this.values[langCode];
+        }
       }
 
```

Comment:
So just to double check - you don't want to include _rm suffix search here for now, in the same way as you did -Latn search?  Because there are four times as many "_rm" suffixes (~260,000) as there are "-Latn" (~75,000), so if anything, we should probably process _rm only.  But _rm tends to be clustered much more in Asia, especially in Korea, whereas -Latn is more global - that's why i think we should do both, and _rm should go after -Latn.

Reply:
hmm... is _rm the same idea as -Latn ? My worry was anything other than _rm, like, generalizing the script, but you're right that in this particular fallback we're using "Latn" anyways, so we might as well?

I would like to have something that also generalizes the process for other potential codes. Is there anything other than _rm and _ro that we should consider for other scripts (possibly in a followup) ?